File: Graphs/DFS.java

import java.util.*;
public class DFS {

    public static void dfsOfaGraph(int start, List<List<Integer>> AdjLs, boolean[] vis, List<Integer> ans)
    {
        //Mark the node a visited
        vis[start]=true;//make it infected.
        //you might get a doubt where won't we check and if it was already true?
        //dfs won't even be called if it is true, dfs is only called if it is false
        ans.add(start);//bus ekkiyyu.

        //traverse all it's neighbours, check for it's friends
        for(int n:AdjLs.get(start)){
            if(vis[n]==false){
              //if friends are not infected call dfs(friend)
                dfsOfaGraph(n,AdjLs,vis,ans);
            }//if they are true means they are already in the bus
        }
    }

}

// Iterative



class Solution {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        ArrayList<ArrayList<Integer>> adj = new ArrayList<ArrayList<Integer>>();
        int noOfVertices = sc.nextInt();
        for (int i = 0; i < noOfVertices; i++) {
            adj.add(new ArrayList<>());
        }

        int noOfEdges = sc.nextInt();

        for (int i = 0; i < noOfEdges; i++) {
            int v1 = sc.nextInt();
            int v2 = sc.nextInt();
            adj.get(v1).add(v2);
            // adj.get(v2).add(v1);
        }

        int src = sc.nextInt();

        List<Integer> result = dfs(src, adj);

        for (int i = 0; i < result.size(); i++) {
            System.out.print(result.get(i) + " ");
        }

    }

    public static List<Integer> dfs(int src, ArrayList<ArrayList<Integer>> adj) {
        Stack<Integer> stk = new Stack<>();
        boolean[] vis = new boolean[adj.size()];// for those many no of edges.
        List<Integer> result = new ArrayList<>();
        List<Integer> temp = new ArrayList<>();
        stk.push(src);
        vis[src] = true;
        while (!stk.isEmpty()) {
            int n = stk.pop();
            result.add(n);
            temp = adj.get(n);
            for (int i = temp.size() - 1; i >= 0; i--) {
                int k = temp.get(i);
                if (!vis[k]) {
                    stk.push(k);
                    vis[k] = true;
                }
            }
        }
        return result;
    }
}

class DFS_Directed_graph_AdjMatrix{

    

}

File: Graphs/DFS.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <queue>
#include <stack>
#include <set>
#include <map>
#include <vector>
using namespace std;

class Solution
{
private:
    void dfs(int node, vector<int> adj[], int vis[], vector<int> &ls)
    {
        vis[node] = 1;
        ls.push_back(node);
        // traverse all its neighbours
        for (auto it : adj[node])
        {
            // if the neighbour is not visited
            if (vis[it] == 0)
            {
                dfs(it, adj, vis, ls);
            }
        }
    }

public:
    // Function to return a list containing the DFS traversal of the graph.
    // it intakes the number of Vertices and the adjacency List
    vector<int> dfsOfGraph(int V, vector<int> adj[])
    {
        vector<int> vis(V, 0); // assuming it starts from 0
        int start = 0;
        // create a list to store dfs
        vector<int> ls;
        // call dfs for starting node
        // dfs(start, adj, vis, ls);
        return ls;
    }
};

void addEdge(vector<int> adj[], int u, int v)
{
    adj[u].push_back(v);
    adj[v].push_back(u);
}

void printAns(vector<int> &ans)
{
    for (int i = 0; i < ans.size(); i++)
    {
        cout << ans[i] << " ";
    }
}

int main()
{
    vector<int> adj[5];

    addEdge(adj, 0, 2);
    addEdge(adj, 2, 4);
    addEdge(adj, 0, 1);
    addEdge(adj, 0, 3);

    Solution obj;
    vector<int> ans = obj.dfsOfGraph(5, adj);
    printAns(ans);

    return 0;
}

// -------------------------------------------------------------------------

// this is a successfull DFS of a graph for any source node.

void Dfs(int node, vector<int> &Vis, vector<int> adj[], vector<int> &dfs)
{
    Vis[node] = 1;
    dfs.push_back(node);
    for (auto it : adj[node])
    {
        if (!Vis[it])
        {
            Dfs(it, Vis, adj, dfs);
        }
    }
}

vector<int> dfsOfGraph(vector<int> adj[], int S, int V)
{
    vector<int> Vis(V, 0); // 0-based indexing
    vector<int> dfs;       // where the answer of dfs will be stored.
    Dfs(S, Vis, adj, dfs);
    return dfs;
}

void addEdge(vector<int> adj[], int n, int m)
{

    adj[n].push_back(m);
    adj[m].push_back(n);
}

void PrintAns(vector<int> ans)
{
    for (int i = 0; i < ans.size(); i++)
    {
        cout << ans[i] << " ";
    }
}

int main()
{
    vector<int> adj[7];
    addEdge(adj, 0, 1);
    addEdge(adj, 1, 2);
    addEdge(adj, 2, 3);
    addEdge(adj, 3, 1);
    addEdge(adj, 3, 4);
    addEdge(adj, 3, 6);
    addEdge(adj, 6, 5);
    addEdge(adj, 4, 5);
    vector<int> dfs = dfsOfGraph(adj, 2, 7);
    PrintAns(dfs);

    return 0;
}